### ISSUE TITLE
Creation of custom sidechainnet enties for .cif files

### ISSUE THREAD

**Issue by alpha-omega-labs** [jonathanking/sidechainnet#53]
Hello,
What is the way to create custom sidechainnet (and proteinnet) alike dataset entries from a list of .cif files?
How to create sidechainnet formatted dataset with 1000 custom .cif proteins from RCSB?
Can sidechainnet entries be used along with OpenProtein https://github.com/biolib/openprotein?
Thank you!

**Comment by alpha-omega-labs** [comment]
Hello,
Could you please provide some heads up on how to better use Sidechainnet for the following goal:

Develop model that predicts some specific proteins class (by structure) - 100-400aa length.
This class is divided into two sub classes, that structurally may be descried as similar as cat and bobcat. 
Model should be able to predict both sub classes accurately.
Proteinnet CASP12 have <150 of chains related to class 1 and <150 chains related to class 2.
Current RCSB have ~800 of class 1 and ~500 of class 2.

Questions:

1. Is it better to develop 2 models - separate for class 1 and another for class 2, OR develop single model for both classes? (so train on 800 and 500 separately OR train on 1300 and have proper splits in validation subset).
2. How to properly construct dataset for that having all: chains from RCSB and also Proteinnet existing records (~300). Is it proper way to load proteinnet with existing IDs (~300) and add non existing chains via provided commands? Or is it better to load one single ID from proteinnet and other 1299 from RCSB? 
3. What model better to use for such structures of ~300aa and with about 1300 items in dataset? (real world data to predict with model is about 35k sequences). 
4. Is there a discord or any other place to hav chat for general questions about sidechainnet?

Thank you

**Comment by jonathanking** [comment]
> Hello,
> What is the way to create custom sidechainnet (and proteinnet) alike dataset entries from a list of .cif files?
> How to create sidechainnet formatted dataset with 1000 custom .cif proteins from RCSB?
> Can sidechainnet entries be used along with OpenProtein https://github.com/biolib/openprotein?
> Thank you!

I'm sorry, but at the moment SidechainNet cannot parse cif files. [This issue](https://github.com/jonathanking/sidechainnet/issues/44#issuecomment-1105700760) shows a way to load PDB files, however. To convert CIF to PDB files, you may try using ProDy. Let me know if you need more assistance with that.

I am unfamiliar with OpenProtein, unfortunately, so I'm not sure how to answer your question.

**Comment by jonathanking** [comment]
First of all, I want to thank you so much for your interest in using SidechainNet! I hope I can help you out here.

> Questions:
> 
> 1. Is it better to develop 2 models - separate for class 1 and another for class 2, OR develop single model for both classes? (so train on 800 and 500 separately OR train on 1300 and have proper splits in validation subset).

I think this is a great research question that I do not have the answer for myself. My personal hunch would be to use a single model.

> 2. How to properly construct dataset for that having all: chains from RCSB and also Proteinnet existing records (~300). Is it proper way to load proteinnet with existing IDs (~300) and add non existing chains via provided commands? Or is it better to load one single ID from proteinnet and other 1299 from RCSB?

I'm sorry, I am not sure what you are suggesting. Could you please clarify? 
A) Are you trying to load proteinnet directly, or use SidechainNet?
B) What do you mean "load one single ID from ProteinNet"? 

If I understand you correctly, I think you are just trying to load these proteins into the SidechainNet format so you can train a model. In that case, I would simply use the custom SidechainNet constructors that are demonstrated in the Colab notebook.

> 3. What model better to use for such structures of ~300aa and with about 1300 items in dataset? (real world data to predict with model is about 35k sequences).

Unfortunately, exactly what model to use is again really hard to suggest. New papers are being released every week or so.

> 4. Is there a discord or any other place to hav chat for general questions about sidechainnet?

No, there is no discord for SidechainNet. Feel free to open discussions here, though!

**Comment by alpha-omega-labs** [comment]
Thank you very much for detailed answers! 

Main issue now is batch shuffle, so here is the setup:

1. There are 800 proteins of approx same length (length may matter for batching probably)
2. 768 in 100% custom "train" set, and copy of it in "valid-10" named set. Both are exact same but produce diff batching.     print("Protein IDs\n   ", batch.pids)
3. If "train" set is used for training (with 768 proteins) and no matter what batch size - its always once single PDBID repeated n batch_size times.    print("Protein IDs\n   ", batch.pids)
4. If "valid-10" set is used for training (with 768 proteins) - its different proteins in single batch, however each batch has same set of protein. (and number seems to be random sometimes)

So despite "train" and "valid-10" are identical it produce diff batches.

<pre>
for epoch in range(1000):
  #print(f"Epoch {epoch}")
  #progress_bar = tqdm(total=len(d['train']), smoothing=0)
  i = 5
  for batch in d['valid-10']:
      i -= 1
      if i <= 0:
          break
      print(f"Model Input = {tuple(batch.seq_evo_sec.shape)}; Total residues = "
          f"{batch.seq_evo_sec.shape[0] * batch.seq_evo_sec.shape[1]}.")	 
      # Prepare variables and create a mask of missing angles (padded with zeros)
      # Note the mask is repeated in the last dimension to match the sin/cos represenation.
      seq_evo_sec = batch.seq_evo_sec.to(device)
      true_angles_sincos = scn.structure.trig_transform(batch.angs).to(device)
      mask = (batch.angs.ne(0)).unsqueeze(-1).repeat(1, 1, 1, 2)
      
      # Make predictions and optimize
      pred_angles_sincos = pssm_model(seq_evo_sec)
      loss = mse_loss(pred_angles_sincos[mask], true_angles_sincos[mask])
      loss.backward()
      torch.nn.utils.clip_grad_norm_(pssm_model.parameters(), 2)
      optimizer.step()

      # Housekeeping
      batch_losses.append(float(loss))
      #progress_bar.update(1)
      #progress_bar.set_description(f"\rRMSE Loss = {np.sqrt(float(loss)):.4f}")
      
      valid_d = d['valid-10']

      valid10_batch = next(iter(valid_d)) 
      print(valid10_batch._fields)	     
      print(valid10_batch.pids)
      print("Protein IDs\n   ", batch.pids)

   </pre>

Could sidechainnet batch loader based on seq. length be not used, especially if proteins in input are approx same size? How to shuffle custom dataset directly based on batch size, e.g. if batch size is 32, how to make batches going 1-32, 32-64, 64-96, etc
what are consequences of much variable chain length in batch vs approx same length?
